File: backend/app/services/pr.py
"""PR detection — does a freshly-logged Set beat the user's history?

CONTEXT.md ("PR"): "A user's personal record for an Exercise — best weight,
reps-at-weight, estimated 1RM, or volume; detected live as a Set is logged
(offline included) and celebrated in the UI."

This is a **pure** module: ``detect_prs`` takes the candidate Set's numbers plus a
snapshot of the user's prior bests for that Exercise and returns the dimensions
beaten. It holds no DB or I/O, so it is:

* the single shared algorithm definition — the backend calls it to compute and
  persist authoritative PRs on sync (record of truth), and a hand-mirrored TS
  port (``frontend/src/lib/pr.ts``) runs the same logic in the browser so PRs
  fire instantly while **offline**, with no server round-trip;
* trivially testable, with the four dimensions, the normal-only rule and the
  strict-improvement rule pinned by :mod:`tests.test_pr`.

The four PR dimensions (CONTEXT.md):

* **weight**          — heaviest weight lifted, at any rep count;
* **e1rm**            — highest estimated 1RM (see :mod:`app.services.e1rm`);
* **reps_at_weight**  — most reps at a *given* weight (keyed per weight);
* **volume**          — biggest single-set volume load (weight × reps).

Rules, all enforced here so callers (and the TS port) inherit them:

* only **normal** Sets qualify — warmup/drop/failure never PR (the exclusion is
  owned by :func:`app.services.volume.counts_for_volume`, reused here, *not*
  re-listed);
* a PR is a **strict** improvement over the prior best — ties are not PRs;
* a Set with zero weight or zero reps never PRs (a 0-load placeholder is noise);
* against an empty history every applicable dimension is a PR (the first time you
  log an Exercise, that Set is your record on every axis).
"""

from __future__ import annotations

import enum
from dataclasses import dataclass, field


class PRKind(str, enum.Enum):
    """The four dimensions a Set can set a personal record on.

    A ``str`` enum (like :class:`~app.models.training_session.SetType`) so it is a
    typed, GROUP-BY-able dimension and serialises to its label on the wire — the
    same labels the TS ``PRKind`` union uses, keeping Py and TS in lockstep.

    Defined *before* the model/service imports below so it is already bound when
    ``app.models.personal_record`` (which keys its native enum column on this
    type) is imported as part of that cascade — otherwise importing this module
    first would deadlock on a partially-initialised ``PRKind`` (circular import).
    """

    weight = "weight"
    e1rm = "e1rm"
    reps_at_weight = "reps_at_weight"
    volume = "volume"


from app.models.training_session import SetType  # noqa: E402
from app.services.e1rm import estimated_1rm  # noqa: E402
from app.services.volume import counts_for_volume  # noqa: E402


@dataclass(frozen=True)
class PriorBests:
    """A snapshot of a user's prior bests for one Exercise (normal Sets only).

    Every field defaults to "nothing logged yet" so an empty ``PriorBests()`` is
    a clean slate on which any qualifying Set is a PR. ``reps_by_weight`` maps an
    exact weight to the most reps ever done at it — the per-weight key for the
    reps-at-weight dimension; a weight absent from the map has never been lifted.
    """

    best_weight_kg: float | None = None
    best_e1rm: float | None = None
    best_volume_kg: float | None = None
    reps_by_weight: dict[float, int] = field(default_factory=dict)


@dataclass(frozen=True)
class PRResult:
    """One PR a Set achieved: the dimension and the value that set it.

    ``at_weight_kg`` is populated only for :attr:`PRKind.reps_at_weight` (the
    weight the rep PR happened at); it is ``None`` for the weight-independent
    dimensions. ``value`` is the achieved figure — the new record — for the UI
    banner ("New 5-rep PR — 100 kg!") and for the persisted record.
    """

    kind: PRKind
    value: float
    at_weight_kg: float | None = None


def _beats(candidate: float, prior: float | None) -> bool:
    """True if ``candidate`` strictly improves on ``prior`` (None = no prior)."""
    return prior is None or candidate > prior


def detect_prs(
    *,
    weight_kg: float,
    reps: int,
    set_type: SetType,
    rir: int | None,
    prior: PriorBests,
) -> list[PRResult]:
    """Return the PRs a just-logged Set sets, given the user's prior bests.

    Keyword-only to keep call sites self-documenting (the weight/reps order is
    easy to flip). Returns an empty list when the Set does not qualify (non-normal
    type, zero load, or zero reps) or beats nothing.
    """
    # Non-normal Sets never PR — single source of the rule is volume.counts_for_volume.
    if not counts_for_volume(set_type):
        return []
    # A zero-load or zero-rep Set carries no signal; never a PR.
    if weight_kg <= 0 or reps <= 0:
        return []

    results: list[PRResult] = []

    # 1. Best weight, at any rep count.
    if _beats(weight_kg, prior.best_weight_kg):
        results.append(PRResult(kind=PRKind.weight, value=weight_kg))

    # 2. Best estimated 1RM (Effort-adjusted via RIR — a set with reserve is
    #    effectively heavier, so it can tip an e1RM PR; see services.e1rm).
    e1rm = estimated_1rm(weight_kg, reps, rir=rir)
    if _beats(e1rm, prior.best_e1rm):
        results.append(PRResult(kind=PRKind.e1rm, value=e1rm))

    # 3. Best reps at THIS exact weight (first ever at a weight always qualifies).
    prior_reps_here = prior.reps_by_weight.get(weight_kg)
    if _beats(reps, prior_reps_here):
        results.append(
            PRResult(
                kind=PRKind.reps_at_weight,
                value=float(reps),
                at_weight_kg=weight_kg,
            )
        )

    # 4. Best single-set volume load (weight × reps).
    volume = weight_kg * reps
    if _beats(volume, prior.best_volume_kg):
        results.append(PRResult(kind=PRKind.volume, value=volume))

    return results

File: frontend/src/lib/pr.ts
/**
 * PR detection + e1RM — the browser-side mirror of the backend pure cores.
 *
 * This is a hand-mirrored port of `backend/app/services/e1rm.py` and
 * `backend/app/services/pr.py`. It runs PR detection in the browser so a
 * personal record fires **instantly while logging — offline included** (the app
 * is offline-first; ADR-0005), with no server round-trip. The backend re-runs
 * the same algorithm on sync as the record-of-truth and reconciles any
 * client-side celebration that turns out to be wrong.
 *
 * The two implementations are kept provably in agreement by mirrored test suites
 * (`pr.test.ts` here, `tests/test_e1rm.py` + `tests/test_pr.py` on the backend) —
 * the same cases, the same expected outcomes. If you change the formula or a PR
 * rule, change it in BOTH and update both suites.
 *
 * The four PR dimensions (CONTEXT.md "PR"):
 *   weight          — heaviest weight, at any reps
 *   e1rm            — highest estimated 1RM
 *   reps_at_weight  — most reps at a given weight (keyed per weight)
 *   volume          — biggest single-set volume load (weight × reps)
 *
 * Rules (identical to the backend): only `normal` sets qualify; a PR is a STRICT
 * improvement (ties don't count); a zero-load or zero-rep set never PRs; against
 * an empty history every applicable dimension is a PR.
 */

import type { SetType } from './types';

/** Epley's denominator constant — one formula, one place (matches e1rm.py). */
const EPLEY_DIVISOR = 30;

/**
 * 1-rep-anchored Epley estimate for `weightKg` lifted for `reps` reps.
 *
 * Uses `(reps - 1)` so a single returns the weight exactly (the boundary PR
 * detection relies on). Returns 0 if either input is 0; throws on negatives —
 * mirroring `epley_1rm` in e1rm.py.
 */
export function epley1rm(weightKg: number, reps: number): number {
  if (weightKg < 0) throw new Error(`weightKg must be >= 0, got ${weightKg}`);
  if (reps < 0) throw new Error(`reps must be >= 0, got ${reps}`);
  if (weightKg === 0 || reps === 0) return 0;
  return weightKg * (1 + (reps - 1) / EPLEY_DIVISOR);
}

/**
 * Estimated 1RM for a set, optionally adjusted for reps in reserve (RIR).
 *
 * A positive RIR makes the set effectively heavier (`effectiveReps = reps +
 * rir`) so the estimate rises and never falls; `rir` of 0 (to failure) or
 * null/undefined (not rated) leaves the plain Epley result. Mirrors
 * `estimated_1rm` in e1rm.py.
 */
export function estimated1rm(
  weightKg: number,
  reps: number,
  rir: number | null = null,
): number {
  if (rir != null && rir < 0) throw new Error(`rir must be >= 0, got ${rir}`);
  const effectiveReps = reps + (rir ?? 0);
  return epley1rm(weightKg, effectiveReps);
}

/** The four dimensions a set can PR on — same labels as the backend `PRKind`. */
export type PRKind = 'weight' | 'e1rm' | 'reps_at_weight' | 'volume';

/**
 * A snapshot of a user's prior bests for one Exercise (normal sets only).
 *
 * Every field is "nothing logged yet" by default, so an empty `PriorBests` is a
 * clean slate. `repsByWeight` maps an exact weight to the most reps ever done at
 * it; a weight absent from it has never been lifted. Mirrors `PriorBests`.
 */
export interface PriorBests {
  bestWeightKg: number | null;
  bestE1rm: number | null;
  bestVolumeKg: number | null;
  repsByWeight: Map<number, number>;
}

/** A clean-slate PriorBests (no history). */
export function emptyPriorBests(): PriorBests {
  return {
    bestWeightKg: null,
    bestE1rm: null,
    bestVolumeKg: null,
    repsByWeight: new Map(),
  };
}

/** One PR a set achieved: the dimension, the new record value, and (for
 * reps-at-weight) the weight it happened at. Mirrors `PRResult`. */
export interface PRResult {
  kind: PRKind;
  value: number;
  atWeightKg: number | null;
}

/** True if `candidate` strictly improves on `prior` (null = no prior). */
function beats(candidate: number, prior: number | null): boolean {
  return prior == null || candidate > prior;
}

/** Whether a set of this type contributes to volume/PR stats (only `normal`).
 * Mirrors `services/volume.py` `counts_for_volume`. */
export function countsForPr(setType: SetType): boolean {
  return setType === 'normal';
}

export interface DetectPrsInput {
  weightKg: number;
  reps: number;
  setType: SetType;
  rir: number | null;
  prior: PriorBests;
}

/**
 * Return the PRs a just-logged set sets, given the user's prior bests.
 *
 * Empty when the set does not qualify (non-normal, zero load, or zero reps) or
 * beats nothing. Byte-for-byte the same logic as `detect_prs` in pr.py.
 */
export function detectPrs({
  weightKg,
  reps,
  setType,
  rir,
  prior,
}: DetectPrsInput): PRResult[] {
  // Non-normal sets never PR (the volume.py exclusion).
  if (!countsForPr(setType)) return [];
  // A zero-load or zero-rep set carries no signal.
  if (weightKg <= 0 || reps <= 0) return [];

  const results: PRResult[] = [];

  // 1. Best weight, at any rep count.
  if (beats(weightKg, prior.bestWeightKg)) {
    results.push({ kind: 'weight', value: weightKg, atWeightKg: null });
  }

  // 2. Best estimated 1RM (Effort-adjusted via RIR).
  const e1rm = estimated1rm(weightKg, reps, rir);
  if (beats(e1rm, prior.bestE1rm)) {
    results.push({ kind: 'e1rm', value: e1rm, atWeightKg: null });
  }

  // 3. Best reps at THIS exact weight (first ever at a weight always qualifies).
  const priorRepsHere = prior.repsByWeight.get(weightKg) ?? null;
  if (beats(reps, priorRepsHere)) {
    results.push({ kind: 'reps_at_weight', value: reps, atWeightKg: weightKg });
  }

  // 4. Best single-set volume load (weight × reps).
  const volume = weightKg * reps;
  if (beats(volume, prior.bestVolumeKg)) {
    results.push({ kind: 'volume', value: volume, atWeightKg: null });
  }

  return results;
}

/**
 * Build a PriorBests from a flat list of a user's prior NORMAL sets for one
 * Exercise (caller filters to the right exercise + normal type, and excludes the
 * candidate set). Mirrors the aggregation in `prior_bests_for` so the offline
 * detector sees the same history shape the backend computes.
 */
export function priorBestsFromSets(
  sets: { weightKg: number; reps: number; rir: number | null }[],
): PriorBests {
  const prior = emptyPriorBests();
  for (const s of sets) {
    if (s.weightKg <= 0 || s.reps <= 0) continue;
    if (prior.bestWeightKg == null || s.weightKg > prior.bestWeightKg) {
      prior.bestWeightKg = s.weightKg;
    }
    const volume = s.weightKg * s.reps;
    if (prior.bestVolumeKg == null || volume > prior.bestVolumeKg) {
      prior.bestVolumeKg = volume;
    }
    const e1rm = estimated1rm(s.weightKg, s.reps, s.rir);
    if (prior.bestE1rm == null || e1rm > prior.bestE1rm) {
      prior.bestE1rm = e1rm;
    }
    const priorReps = prior.repsByWeight.get(s.weightKg) ?? null;
    if (priorReps == null || s.reps > priorReps) {
      prior.repsByWeight.set(s.weightKg, s.reps);
    }
  }
  return prior;
}

/**
 * Human banner copy for a PR, mobile-first and non-blocking (CONTEXT.md: "New
 * 5-rep PR — 100 kg!"). One short phrase per dimension.
 */
export function prLabel(pr: PRResult): string {
  switch (pr.kind) {
    case 'weight':
      return `New weight PR — ${formatKg(pr.value)} kg!`;
    case 'e1rm':
      return `New estimated 1RM — ${formatKg(pr.value)} kg!`;
    case 'reps_at_weight':
      return `New ${pr.value}-rep PR — ${formatKg(pr.atWeightKg ?? 0)} kg!`;
    case 'volume':
      return `New volume PR — ${formatKg(pr.value)} kg!`;
  }
}

/** Trim a kg value to at most one decimal, dropping a trailing ".0". */
function formatKg(n: number): string {
  return Number.isInteger(n) ? String(n) : n.toFixed(1);
}
